runsc/fsgofer/lisafs.go

```diff
@@ -823,6 +839,34 @@ func (fd *controlFDLisa) Connect(sockType uint32) (int, error) {
 	return sock, nil
 }
 
+// ConnectWithCreds implements lisafs.ControlFDImpl.ConnectWithCreds.
+func (fd *controlFDLisa) ConnectWithCreds(sockType uint32, uid lisafs.UID, gid lisafs.GID) (int, error) {
+	serverConfig := fd.Conn().ServerImpl().(*LisafsServer).config
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	_, _, err := unix.Syscall(unix.SYS_SETREUID, uintptr(serverConfig.RUID), uintptr(uid), 0)
+	if !errors.Is(err, syscall.Errno(0)) {
+		log.Warningf("failed to set euid; err: %v", err)
+	}
```

Comment:
I am not sure how to retain the capabilities considering `setreuid(2)` drops them. I saw the option of using `PR_SET_KEEPCAPS` but that only retains the permitted set but clears the effective set. And I see https://github.com/google/gvisor/blob/0fb4cc71ba95bdc1e12ae87ff81476207a73b052/runsc/cmd/cmd.go#L82

re-applies the capabilities but then I need to restart the process which is not suitable for this case. Do you have any advice?

Reply:
I believe this is not a concern as the effective set is restored to the permitted set after restoring the euid.